File: scripts/BDDSynthesisScript.lua
-- BDDSynthesizerScript.lua, Lua script example of how to use the BDD syntehsizer in Lua
-- Meant to be run as a script inside Supremica (with LuaJ embedded)
local luaj = luajava -- just shorthand 
local script, ide = ... -- grab the arguments passed from Java via LuaJ

-- Get convenience function to generate file name
local Config = luaj.bindClass("org.supremica.properties.Config")
local getFileName = dofile(Config.FILE_SCRIPT_PATH:getValue():getPath().."/getFileName.lua")

-- Set up some helpers
local actions = ide:getActions()
local editorSynthesizerAction  = actions.editorSynthesizerAction
local manager = ide:getDocumentContainerManager() 

-- bindClass is like Java's import
local EventKind = luaj.bindClass("net.sourceforge.waters.model.base.EventKind")
local SynthesisType = luaj.bindClass("org.supremica.automata.algorithms.SynthesisType")
local SynthesisAlgorithm = luaj.bindClass("org.supremica.automata.algorithms.SynthesisAlgorithm")

-- Set the options
local options = luaj.newInstance("org.supremica.automata.algorithms.EditorSynthesizerOptions")
options:setSaveInFile(false)
options:setPrintGuard(false)
options:setSynthesisType(SynthesisType.NONBLOCKING_CONTROLLABLE)
options:setSynthesisAlgorithm(SynthesisAlgorithm.MONOLITHICBDD) 
-- options:setSynthesisAlgorithm(SynthesisAlgorithm.PARTITIONBDD)
options:setExpressionType(options.ExpressionType.ADAPTIVE)
options:setAddGuards(false)

-- For all open modules, perform synthesis
local recent = manager:getRecent() -- java.util.List<DocumentContainer> of all open modules
for i = 1, recent:size() do
	local container = recent:get(i-1)
	local name = container:getName()
	if name ~= "New Module" then 
		print("**** Synthesizing for: "..name)
		local module = container:getEditorPanel():getModuleSubject()

		-- collect the controllable events, only them can we assign guards to
		local controllableEvents = luaj.newInstance("java.util.Vector")
		local eventDeclList = module:getEventDeclListModifiable()
		for i = 1, eventDeclList:size() do
		  local event = eventDeclList:get(i-1)
		  if event:getKind() == EventKind.CONTROLLABLE then
			controllableEvents:add(event:getName())
		  end
		end

		local exAutomata = luaj.newInstance("org.supremica.automata.ExtendedAutomata", module)
		local bddSynthesizer = luaj.newInstance("org.supremica.automata.BDD.EFA.BDDExtendedSynthesizer", exAutomata, options)
		bddSynthesizer:synthesize(options)
		bddSynthesizer:generateGuard(controllableEvents, options)

		local saveFile = luaj.newInstance("java.io.File", getFileName("BDDoutput.txt"))
		editorSynthesizerAction:saveOrPrintGuards(bddSynthesizer, controllableEvents,
											options:getSaveInFile(), options:getPrintGuard(),
											saveFile)

		print("Synthesis took "..bddSynthesizer:getSynthesisTimer():toString())
		print("Guards generated in "..bddSynthesizer:getGuardTimer():toString())
		
		if options:getAddGuards() then
		  bddSynthesizer:addGuardsToAutomata()
		end

		-- Cleanup...
		bddSynthesizer:done()
	end
end

File: scripts/BDDSynthesisScript.java
// How to synthesize using the BDD synthesizer
// The options are set in an instance of org\supremica\automata\algorithms\EditorSynthesizerOptions.java
// The options seem to be remembered, so need only be set up once, can be done manually
// Code below is extracted from src\org\supremica\gui\ide\actions\EditorSynthesizerAction.java
// This code can be run by the Editor > Run script...
// Use CreateOpenModule.java as template
package Lupremica;

import java.util.Vector;

import org.supremica.gui.ide.IDE;
import org.supremica.gui.ide.actions.Actions;
import org.supremica.automata.ExtendedAutomata;
import org.supremica.automata.BDD.EFA.BDDExtendedSynthesizer;
import org.supremica.gui.ide.actions.EditorSynthesizerAction;
import org.supremica.automata.algorithms.EditorSynthesizerOptions;
import net.sourceforge.waters.subject.module.EventDeclSubject;
import net.sourceforge.waters.subject.module.ModuleSubject;
import net.sourceforge.waters.model.base.EventKind;
import org.supremica.automata.algorithms.SynthesisType;
import org.supremica.automata.algorithms.SynthesisAlgorithm;
import org.apache.logging.log4j.Logger;

public class BDDSynthesisScript
{
    private final IDE ide;

    public BDDSynthesisScript(final IDE ide)
    {
        this.ide = ide;

		final Logger logger = ide.getTheLog();

		// Look up the editorSynthesizerAction
		final Actions actions = ide.getActions();
		EditorSynthesizerAction editorSynthesizerAction =
				(EditorSynthesizerAction)actions.editorSynthesizerAction;

        final ModuleSubject module =
              ide.getActiveDocumentContainer().getEditorPanel().getModuleSubject();

    	// get the stored or default options
    	final EditorSynthesizerOptions options = new EditorSynthesizerOptions();
		options.setSaveInFile(true);
		options.setPrintGuard(true);
		options.setSynthesisType(SynthesisType.NONBLOCKING_CONTROLLABLE);
		options.setSynthesisAlgorithm(SynthesisAlgorithm.PARTITIONBDD);

        // collect controllable event names from the module
        final Vector<String> controllableEventNames = new Vector<String>();
        for (final EventDeclSubject sigmaS : module.getEventDeclListModifiable())
        {
            if (sigmaS.getKind() == EventKind.CONTROLLABLE)
            {
                controllableEventNames.add(sigmaS.getName());
            }
        }

        final ExtendedAutomata exAutomata = new ExtendedAutomata(module);

        BDDExtendedSynthesizer bddSynthesizer = null;
        RuntimeException synthesisException = null;

        try
        {
            bddSynthesizer = new BDDExtendedSynthesizer(exAutomata, options);
            bddSynthesizer.synthesize(options);
            bddSynthesizer.generateGuard(controllableEventNames, options);
            final java.io.File saveFile = new java.io.File("R:/BDDsynthOutput.txt");
            editorSynthesizerAction.saveOrPrintGuards(bddSynthesizer, controllableEventNames,
                                    options.getSaveInFile(), options.getPrintGuard(),
                                    saveFile);

            if (options.getAddGuards())
                bddSynthesizer.addGuardsToAutomata();

            // Cleanup...
            bddSynthesizer.done();
            bddSynthesizer = null;
        }
        catch (final RuntimeException e)
        {
            synthesisException = e;
        }
        catch(final OutOfMemoryError e)
        {
			synthesisException = new RuntimeException(e.getCause());
			logger.error("BDDSynthesisScript out of memory during BDD synthesis "
			          + " (" + Runtime.getRuntime().maxMemory()/(1024*1024) + " Mib)");
      		e.printStackTrace();
		}
		finally
        {
            if (bddSynthesizer != null)
                bddSynthesizer.done();

            if (synthesisException != null)
                throw synthesisException;
        }
    }
}